### PR TITLE
Gitlab

### PR DESCRIPTION
**Pull request by ankitwagh617**

### CODE REVIEW

File: cartography/intel/gitlab/group.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+from cartography.util import make_requests_url
+from cartography.util import run_cleanup_job
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+
+@timeit
+def get_groups(access_token: str):
+    url = f"https://gitlab.example.com/api/v4/groups?pagelen=100"
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/group.py:18]: is this the gitlab api endpoint? gitlab.example.com? can you add the documentation url in the source code as a comment. would be helpful for future readability and maintenance.

File: cartography/intel/gitlab/members.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+from cartography.util import make_requests_url
+from cartography.util import run_cleanup_job
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+
+@timeit
+def get_group_members(access_token:str,group:str):
+    url = f"https://gitlab.example.com/api/v4/groups/{group}/members?per_page=100"
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/members.py:18]: please add a comment in the source code to show the documentation url of gitlab

File: cartography/intel/gitlab/members.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    """
+    tx.run(
+        ingest_group,
+        membersData=members_data,
+        UpdateTag=common_job_parameters['UPDATE_TAG'],
+    )
+
+
+def cleanup(neo4j_session: neo4j.Session, common_job_parameters: Dict) -> None:
+    run_cleanup_job('gitlab_group_member_cleanup.json', neo4j_session, common_job_parameters)
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/members.py:61]: I don't see the cleanup file in the PR

File: cartography/intel/gitlab/projects.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+from cartography.util import make_requests_url
+from cartography.util import run_cleanup_job
+from cartography.util import timeit
+logger = logging.getLogger(__name__)
+
+
+@timeit
+def get_projects(access_token:str,group:str):
+    url = f"https://gitlab.example.com/api/v4/groups/{group}/projects?per_page=100"
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/projects.py:17]: same, add documentation url in the source code as a comment.

File: cartography/intel/gitlab/projects.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    tx.run(
+        ingest_group,
+        projectData=project_data,
+        UpdateTag=common_job_parameters['UPDATE_TAG'],
+    )
+
+
+def cleanup(neo4j_session: neo4j.Session,  common_job_parameters: Dict) -> None:
+    run_cleanup_job('gitlab_group_project_cleanup.json', neo4j_session, common_job_parameters)
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/projects.py:65]: cleanup file missing from PR

File: cartography/intel/gitlab/resources.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from typing import Dict
+
+from . import members
+from . import projects
+from . import repositories
+
+
+RESOURCE_FUNCTIONS: Dict = {
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/resources.py:8]: I don't see repository. let's add repository as well. That's a key resource for gitlab inventory

**Review by mpurusottamc — CHANGES_REQUESTED** [review]
there are a few changes needed.

File: cartography/intel/gitlab/__init__.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                        config,
+                        group_name,
+                        access_token,
+                        common_job_parameters,
+
+                    ),
+                )
+            else:
+                raise ValueError(
+                    f'Azure sync function "{request}" was specified but does not exist. Did you misspell it?',
```
[the remark this answers is not in the document]
Comment (ankitwagh617) [reply cartography/intel/gitlab/__init__.py]: Done

File: cartography/intel/gitlab/__init__.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    neo4j_session: neo4j.Session,
+    access_token: str,
+    groups:List[Dict],
+    common_job_parameters: Dict[str, Any],
+    config: Config,
+) ->bool:
+    for group in groups:
+        common_job_parameters['GROUP_ID']=group.get('id')
+        _sync_one_gitlab_group(neo4j_session,group.get('path'),access_token,common_job_parameters,config)
+        run_cleanup_job('gitlab_group_cleanup.json', neo4j_session, common_job_parameters)
```
[the remark this answers is not in the document]
Comment (ankitwagh617) [reply cartography/intel/gitlab/__init__.py:84]: Added

File: cartography/intel/gitlab/group.py (hunk 1 above)
Comment (ankitwagh617) [reply cartography/intel/gitlab/group.py:21]: Yes, added

File: cartography/intel/gitlab/members.py (hunk 2 above)
Comment (ankitwagh617) [reply cartography/intel/gitlab/members.py:21]: yes,added

File: cartography/intel/gitlab/members.py (hunk 3 above)
Comment (ankitwagh617) [reply cartography/intel/gitlab/members.py:64]: Added

File: cartography/intel/gitlab/projects.py (hunk 4 above)
Comment (ankitwagh617) [reply cartography/intel/gitlab/projects.py]: yes, added

File: cartography/intel/gitlab/resources.py (hunk 6 above)
Comment (ankitwagh617) [reply cartography/intel/gitlab/resources.py:9]: done

File: cartography/intel/gitlab/dependencies.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    SET dep.name = dependency.name,
+    dep.version = dependency.version,
+    dep.type = dependency.type,
+    dep.description = dependency.description,
+    dep.lastupdated = $UpdateTag
+
+    WITH dep, dependency
+    MATCH (repo:GitLabRepository {id: dependency.repository_id})
+    MERGE (dep)-[r:DEPENDS_ON]->(repo)
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/dependencies.py:52]: we should change this relationsip. it should say repository -> `HAS` -> Dependency

File: cartography/intel/gitlab/repositories.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    re.web_url = repo.web_url,
+    re.last_activity_at = repo.last_activity_at,
+    re.owner_username = repo.owner.username,
+    re.owner_id = repo.owner.id,
+    re.namespace_id = repo.namespace.id,
+    re.namespace_name = repo.namespace.name,
+    re.lastupdated = $UpdateTag
+    WITH re, repo
+    MATCH (project:GitLabProject{id: repo.namespace.id})
+    MERGE (project)<-[o:HAS_REPOSITORY]-(re)
```
Comment (mpurusottamc) [inline cartography/intel/gitlab/repositories.py:60]: rename this relationship to `HAS`

**Review by mpurusottamc — CHANGES_REQUESTED** [review]
few minor change needed. @ganeshjagdale9545 can you review also and help with unit testing?

File: cartography/intel/gitlab/dependencies.py (hunk 9 above)
Comment (ankitwagh617) [reply cartography/intel/gitlab/dependencies.py]: Done

File: cartography/data/jobs/cleanup/gitlab_group_cleanup.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
+	"statements": [
+		{
+			"query": "MATCH (n:GitLabGroup{id: $WORKSPACE_ID})<-[r:OWNER]-(:CloudanixWorkspace{id: $WORKSPACE_ID})  WHERE n.lastupdated <> $UPDATE_TAG WITH n LIMIT $LIMIT_SIZE DETACH DELETE (n) return COUNT(*) as TotalCompleted",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/cleanup/gitlab_group_cleanup.json:4]: use GROUP_ID for GitLabGroup

File: cartography/data/jobs/cleanup/gitlab_group_cleanup.json [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+			"iterative": true,
+			"iterationsize": 100
+		},
+		{
+			"query": "MATCH (:GitLabGroup{id: $WORKSPACE_ID})<-[r:OWNER]-(:CloudanixWorkspace{id: $WORKSPACE_ID}) WHERE r.lastupdated <> $UPDATE_TAG WITH r LIMIT $LIMIT_SIZE DELETE (r) return COUNT(*) as TotalCompleted",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/cleanup/gitlab_group_cleanup.json:9]: same as above

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
Minor change needed
In cleanup job use GROUP_ID for GitLabGroup

File: cartography/data/jobs/cleanup/gitlab_group_cleanup.json (hunk 12 above)
Comment (ankitwagh617) [reply cartography/data/jobs/cleanup/gitlab_group_cleanup.json]: Done

File: tests/integration/cartography/intel/gitlab/test_members.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+def test_load_member(neo4j_session):
+    cartography.intel.gitlab.members.load_members_data(
+        neo4j_session,
+        tests.data.gitlab.group.MEMBERS,
+        common_job_parameters,
+    )
+    nodes = neo4j_session.run(
+        """
+        MATCH (g:BitbucketMember) RETURN g.name;
```
Comment (mpurusottamc) [inline tests/integration/cartography/intel/gitlab/test_members.py:15]: it says bitbucket member. can you please fix this?

**Review by mpurusottamc — CHANGES_REQUESTED** [review]
one change needed

File: tests/integration/cartography/intel/gitlab/test_members.py (hunk 13 above)
Comment (ankitwagh617) [reply tests/integration/cartography/intel/gitlab/test_members.py]: Done

**Pull request merged**